pkg/ipam/node.go

```diff
@@ -837,13 +840,22 @@ func adjustPreAllocateIfNeeded(node *v2.CiliumNode) {
 	// need to adjust it below.
 	node.Spec.IPAM.PreAllocate = defaults.IPAMPreAllocation
 
-	if lim, ok := limits.Get(node.Spec.ENI.InstanceType); ok {
-		max := lim.Adapters * lim.IPv4
-		if node.Spec.IPAM.PreAllocate > max {
-			node.Spec.IPAM.PreAllocate = max
+	lim, ok := limits.Get(node.Spec.ENI.InstanceType)
+	if !ok {
+		return fmt.Errorf(
+			`unable to find the limits for this instance type '%s',
+			consider setting --update-ec2-adapter-limit-via-api=true`,
+			node.Spec.ENI.InstanceType,
+		)
```

Comment:
At this point in the code, this error can be triggered many times because the caller of this function will retry. Given that, it would be nicer if we used a `sync.Once` to log this only once, as to not pollute the logs. We can also provide the recommendation of setting the specific flag, rather than embedding the mitigation in the error msg. The error returned should simply be `errLimitsNotFound` or whatever the sentinel error variable is called.

We should also consider whether it makes more sense for the caller to log once or whether to do it in this function itself.

Reply:
absolutely, I wasn't too sure about this part anyway. I believe it is out of scope for the PR so I will revert it for now.

Reply:
done 👍